`graph`: prevent unwanted matches in certain URLs

claui opened the pull request:
Some small helper packages on the AUR are entirely self-hosted, i.e. they have no dedicated upstream URL outside the AUR.

Those packages deliver everything they need in and alongside the PKGBUILD, so authors of these packages may choose to *self-reference* their own AUR page in the `url` field.

Examples for such packages with self-referential `url` fields are the pkgbases [what-is-electron](https://aur.archlinux.org/cgit/aur.git/tree/.SRCINFO?h=what-is-electron) and [electron-blur-me-not](https://aur.archlinux.org/cgit/aur.git/tree/.SRCINFO?h=electron-blur-me-not):

```properties
pkgbase = what-is-electron
    pkgdesc = Point /usr/bin/electron to any Electron version you like
    pkgver = 0.1.0
    pkgrel = 1
    url = https://aur.archlinux.org/pkgbase/what-is-electron
```

Because those `url` fields may contain the word `pkgbase` themselves, `aur graph` may erroneously pick those URLs from `.SRCINFO`:
```properties
#                                   ↓↓↓↓↓↓↓
    url = https://aur.archlinux.org/pkgbase/what-is-electron
```

This produces an extraneous nonsense line in the output of `aur graph`, preventing users from installing such packages via `aur sync`.

This PR anchors the `pkgbase` and `pkgname` regular expressions to the beginning of the line in order to prevent this issue.

Full disclosure: I am the maintainer of the mentioned pkgbases.

Earnestly commented:
It is often incorrect to use `/regexp/` as it tends to have the capacity to match far more than is intended.

Could you instead change all of these tests to use direct field matching: `$1 == "pkgbase"`?

This would also obviate the need to use `\t` as the default `FS` value of `<space>` has special meaning to awk:

```
If FS is <space>, skip leading and trailing <blank> and <newline> characters; fields shall be delimited by sets of one or more <blank> or <newline> characters.
```

Here `<blank>` includes horizontal tabs.

Earnestly commented:
Actually before you do anything, I'll write something up. There seems to be a lot of low hanging fruit here. Uno momento

claui commented:
@Earnestly Thanks a lot for your feedback!

How about simplifying `option`:

```awk
function option(opt) {
    if(!(length(opt) && opt == "1"))
        return 1
}
```

like this:

```awk
function option(opt) {
    return opt != 1
}
```

My test cases suggest that both functions should be equivalent:

<details><summary>Test cases</summary>
<pre>BEGIN {
    printf option() == 1
    printf option("") == 1
    printf option(foo) == 1
    printf option(null) == 1
    printf option(0) == 1
    printf option("0") == 1
    printf option(1) == 0
    printf option("1") == 0
    printf option(-1) == 1
    printf option("-1") == 1
    printf option(2) == 1
    printf option("2") == 1
}
</pre>
</details>

claui commented:
Is the `option` function correctly implemented?

The original expression was:

```awk
!length(FOO) || FOO == 1   # !A || B
```

The new `option` implementation would resolve to:

```awk
!(length(FOO) && FOO == "1")   # !(A && B)
```

However, `!A || B` is not equivalent to `!(A && B)`.

The following test cases illustrate this regression:

<details><summary>Current version (passing)</summary>

```awk
#!/usr/bin/awk -v BAR=0 -v BAZ=1 -f

function option(opt) {
    return !length(opt) || opt == 1
}

BEGIN {
    # FOO is unset, so implicitly 1
    if (!option(FOO)) exit(12)
    # BAR is explicitly set to 0
    if (option(BAR)) exit(13)
    # BAZ is explicitly set to 1
    if (!option(BAZ)) exit(14)
}
```
</details> 

<details><summary>New implementation (failling)</summary>

```awk
#!/usr/bin/awk -v BAR=0 -v BAZ=1 -f

function option(opt) {
    return !(length(opt) && opt == "1")
}

BEGIN {
    # FOO is unset, so implicitly 1
    if (!option(FOO)) exit(12)
    # BAR is explicitly set to 0
    if (option(BAR)) exit(13)
    # BAZ is explicitly set to 1
    if (!option(BAZ)) exit(14)
}
```
</details>

claui commented:
Another issue with the proposed refactoring is that it doesn’t take into account that `DEPENDS` and `MAKEDEPENDS` should default to 1, while `OPTDEPENDS` and `PRINTALL` should default to 0.

The current implementations is not entirely consistent to what the manpage says; however, the proposed refactoring is not consistent to either:

| Header | What the manpage says | Current implementation | Proposed refactoring |
|--------|--------|--------|--------|
| `DEPENDS` | Defaults to 1 | Defaults to 1 | Defaults to 1 |
| `MAKEDEPENDS` | Defaults to 1 | Defaults to 1 | Defaults to 1 |
| `CHECKDEPENDS` | Defaults to 0 | Defaults to 1 | Defaults to 1 |
| `OPTDEPENDS` | Not documented | Defaults to 0 | Defaults to 1 |
| `PRINTALL` | Not documented | Defaults to 0 | Defaults to 1 |

@Earnestly Would you prefer to update your proposal here, or rather file a separate pull request due to the complexity involved?

Earnestly commented:
Oh yes, the `option` function et al. is completely up for improvements; it's good of you to have looked into that detail as well as I haven't.

I was going over some iterations on IRC and currently have in mind the following:

```diff
diff --git a/lib/aur-graph b/lib/aur-graph
index 2ec6d60c..3b46044b 100755
--- a/lib/aur-graph
+++ b/lib/aur-graph
@@ -33,52 +33,54 @@ function get_vercmp(ver1, ver2, op) {
     }
 }

-/pkgbase/ {
+function option(opt) {
+    if(!(length(opt) && opt == "1"))
+        return 1
+}
+
+BEGIN {
+    attrs["pkgver"]
+    attrs["makedepends"]
+    attrs["checkdepends"]
+    attrs["optdepends"]
+    attrs["depends"]
+    attrs["provides"]
+}
+
+
+$1 == "pkgbase" {
     in_split_pkg = 0
     pkgbase = $3
     pkgver  = ""

     # track both the pkgbases themselves and their number of deps
     dep_counts[pkgbase] = 0
+    next
 }
 
-/^\tpkgver/ {
-    if (!in_split_pkg) {
+$1 in attrs && !in_split_pkg {
+    if ($1 == "pkgver")
         pkgver = $3
-    }
-}
-
-/^\tdepends/ {
-    if ((!length(DEPENDS) || DEPENDS == 1) && !in_split_pkg)
+    else if ($1 == "depends" && option(DEPENDS))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3 # versioned
-}
-
-/^\tmakedepends/ {
-    if ((!length(MAKEDEPENDS) || MAKEDEPENDS == 1) && !in_split_pkg)
+    else if ($1 == "makedepends" && option(MAKEDEPENDS))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
-}
-
-/^\tcheckdepends/ {
-    if ((!length(CHECKDEPENDS) || CHECKDEPENDS == 1) && !in_split_pkg)
+    else if ($1 == "checkdepends" && option(CHECKDEPENDS))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
-}
-
-/^\toptdepends/ {
-    if (OPTDEPENDS == 1 && !in_split_pkg) {
+    else if ($1 == "optdepends" && option(OPTDEPENDS))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
-    }
-}
 
-/^$/ {
-    in_split_pkg = 1
+    next
 }

-/pkgname/ {
+$1 == "pkgname" {
     pkg_map[$3] = pkgbase # node
     ver_map[$3] = pkgver  # weight
+    next
 }
 
-/^\tprovides/ {
+
+$1 == "provides" {
     split($3, prov, "=")

     # if provider is unversioned, use pkgver
@@ -89,6 +91,12 @@ function get_vercmp(ver1, ver2, op) {

     # append node
     pkg_map[prov[1]] = pkgbase
+    next
+}
+
+!NF {
+    in_split_pkg = 1
+    next
 }

 END {
@@ -99,7 +107,7 @@ END {
         printf("%s\t%s\n", pkgbase, pkgbase)

         for (dep = 1; dep <= dep_counts[pkgbase]; dep++) {
-            if (length(PRINTALL) || PRINTALL == 1) {
+            if (option(PRINTALL)) {
                 printf("%s\t%s\n", pkgbase, pkg_deps[pkgbase, dep])
                 continue
             }
@@ -146,6 +154,5 @@ END {
         }
     }

-    if (_vercmp_exit)
-        exit _vercmp_exit
+    exit _vercmp_exit
 }
```

Now this doesn't include the probably necessary changes to how `OPTDEPENDS` and `PRINTALL` is handled, I haven't given that much thought.

I personally don't mind if you want to take these changes and run with them in your PR. Will have to see what @alad thinks.

Earnestly commented:
Just to keep you up to date, current idea is:

```awk
# ...

function setopt(option, default) {
    return option ~ /[01]/ ? option : default
}

BEGIN {
    dependency["depends"]
    dependency["makedepends"]
    dependency["checkdepends"]
    dependency["optdepends"]

    options["depends"] = setopt(DEPENDS, 1)
    options["makedepends"] = setopt(MAKEDEPENDS, 1)
    options["checkdepends"] = setopt(CHECKDEPENDS, 1)
    options["optdepends"] = setopt(OPTDEPENDS, 0)
    options["printall"] = setopt(PRINTALL, 1)
}

# ...

$1 in dependency && !in_split_pkg {
    if(options[$1])
        pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3 # versioned
    next
}

# ...
```

This also collapses the repeatition of `pkg_deps` assignments.

claui commented:
@Earnestly Thank you for the revised proposal. This looks correct, except for the expression `setopt(PRINTALL, 1)`, which I corrected to `setopt(PRINTALL, 0)`.

Running a final thorough regression test over the 168 packages I happen to have in my `~/.cache/aurutils/sync` directory has uncovered one more regression:

in `.SRCINFO`, the `depends` keyword may be followed by an underscore and an architecture specifier as a suffix, e.g. `depends_x86_64`. The same goes for `checkdepends`, `makedepends`, and `optdepends`.
See `man PKGBUILD` for a description of this feature.

One example of a `.SRCINFO` which includes such keywords is the [`rnix-lsp` package](https://aur.archlinux.org/cgit/aur.git/tree/.SRCINFO?h=rnix-lsp). The original implementation would include the line:

```plain
rnix-lsp	gcc-libs
```

when run with `-v PRINTALL=1`, while your proposal would omit that line.
Aside from that regression, everything looks perfect. Nice work!

For your reference, this is the stress regression test I used:

<details><summary>Details</summary>

```bash
#!/bin/bash
set -eux -o pipefail
cd "$(mktemp -d)"
set +x
find ~/.cache/aurutils/sync -maxdepth 2 -name '.SRCINFO' -exec bash -c '
  srcinfo="${1}"
  id="$(basename "$(dirname "${srcinfo}")")"
  mkdir "${id}"
  set -- awk -f path/to/aurutils/lib/aur-graph
  "$@" "${srcinfo}" > "${id}/00-default.stdout.txt"
  "$@" -v PRINTALL=1 \
    "${srcinfo}" > "${id}/01-printall.stdout.txt"
  "$@" -v DEPENDS=1 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/03-depends.stdout.txt"
  "$@" -v MAKEDEPENDS=1 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/04-makedepends.stdout.txt"
  "$@" -v CHECKDEPENDS=1 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/05-checkdepends.stdout.txt"
  "$@" -v OPTDEPENDS=1 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/06-optdepends.stdout.txt"
  "$@" -v DEPENDS=0 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/07-no-depends.stdout.txt"
  "$@" -v MAKEDEPENDS=0 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/08-no-makedepends.stdout.txt"
  "$@" -v CHECKDEPENDS=0 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/09-no-checkdepends.stdout.txt"
  "$@" -v OPTDEPENDS=0 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/10-no-optdepends.stdout.txt"
  "$@" -v DEPENDS=1 -v MAKEDEPENDS=1 -v CHECKDEPENDS=1 \
    -v OPTDEPENDS=1 -v PRINTALL=1 \
    "${srcinfo}" > "${id}/11-all-options.stdout.txt"
  ' _ '{}' ';'
```
</details>

I ran the stress test once with the original implementation and once with your revised proposal. I then compared the two output sets like so:

```sh
diffoscope --no-default-limits --progress --exclude-directory-metadata=yes \
  /tmp/aur-graph-original \
  /tmp/aur-graph-proposed
```

(Note: This command line requires the `diffoscope` and `python-progressbar` packages.)

Would you mind revising your proposal so it fixes this one last regression?

Earnestly commented:
I had completely forgotten that pacman (makepkg) added `_$CARCH` suffixed versions of `depends` and co. That might be worth investigating in a separate PR as `aur graph` hasn't changed since 2018 or so.

This PR might as well just focus on fixing the immediate issue with a little staging to help address the next.

Earnestly commented:
@claui Sorry for delay, what do you think of this?

```diff
diff --git a/lib/aur-graph b/lib/aur-graph
index 2ec6d60c..9ed9fc16 100755
--- a/lib/aur-graph
+++ b/lib/aur-graph
@@ -1,4 +1,4 @@
-#!/bin/awk -f
+#!/bin/gawk -f
 # aur-graph - print package/dependency pairs
 
 # < 0 : if ver1 < ver2
@@ -33,7 +33,33 @@ function get_vercmp(ver1, ver2, op) {
     }
 }
 
-/pkgbase/ {
+function setopt(opt, default) {
+    if(opt ~ /[01]/)
+       return opt
+    else
+       return default
+}
+
+function has_valid_arch(attr,    i, carch) {
+    if(i = index($1, "_")){
+        carch = substr($1, i+1)
+
+        if(!(carch in arch))
+            return 0
+    }
+
+    return 1
+}
+
+BEGIN {
+    DEPENDS = setopt(DEPENDS, 1)
+    MAKEDEPENDS = setopt(MAKEDEPENDS, 1)
+    CHECKDEPENDS = setopt(CHECKDEPENDS, 1)
+    OPTDEPENDS = setopt(OPTDEPENDS, 0)
+    PRINTALL = setopt(PRINTALL, 1)
+}
+
+$1 == "pkgbase" {
     in_split_pkg = 0
     pkgbase = $3
     pkgver  = ""
@@ -42,43 +68,44 @@ function get_vercmp(ver1, ver2, op) {
     dep_counts[pkgbase] = 0
 }
 
-/^\tpkgver/ {
-    if (!in_split_pkg) {
-        pkgver = $3
-    }
+$1 == "pkgver" && !in_split_pkg {
+    pkgver = $3
 }
 
-/^\tdepends/ {
-    if ((!length(DEPENDS) || DEPENDS == 1) && !in_split_pkg)
-        pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3 # versioned
+$1 == "arch" {
+    arch[$3]
 }
 
-/^\tmakedepends/ {
-    if ((!length(MAKEDEPENDS) || MAKEDEPENDS == 1) && !in_split_pkg)
+index($1, "depends") == 1 && !in_split_pkg {
+    if (DEPENDS && has_valid_arch($1))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
 }
 
-/^\tcheckdepends/ {
-    if ((!length(CHECKDEPENDS) || CHECKDEPENDS == 1) && !in_split_pkg)
+index($1, "makedepends") == 1 && !in_split_pkg {
+    if (MAKEDEPENDS && has_valid_arch($1))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
 }
 
-/^\toptdepends/ {
-    if (OPTDEPENDS == 1 && !in_split_pkg) {
+index($1, "checkdepends") == 1 && !in_split_pkg {
+    if (CHECKDEPENDS && has_valid_arch($1))
+        pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
+}
+
+index($1, "optdepends") == 1 && !in_split_pkg {
+    if (OPTDEPENDS && has_valid_arch($1))
         pkg_deps[pkgbase, ++dep_counts[pkgbase]] = $3
-    }
 }
 
-/^$/ {
+NF == 0 {
     in_split_pkg = 1
 }
 
-/pkgname/ {
+$1 == "pkgname" {
     pkg_map[$3] = pkgbase # node
     ver_map[$3] = pkgver  # weight
 }
 
-/^\tprovides/ {
+$1 ~ /^provides/ && has_valid_arch($1) {
     split($3, prov, "=")
 
     # if provider is unversioned, use pkgver
@@ -99,7 +126,7 @@ END {
         printf("%s\t%s\n", pkgbase, pkgbase)
 
         for (dep = 1; dep <= dep_counts[pkgbase]; dep++) {
-            if (length(PRINTALL) || PRINTALL == 1) {
+            if (PRINTALL) {
                 printf("%s\t%s\n", pkgbase, pkg_deps[pkgbase, dep])
                 continue
             }
@@ -146,6 +173,5 @@ END {
         }
     }
 
-    if (_vercmp_exit)
-        exit _vercmp_exit
+    exit _vercmp_exit
 }
```

claui commented:
@Earnestly There was an issue with operator precedence in your `has_valid_arch` function, which prevented your fix from taking effect. (Variable assignment is weaker than `&&`.)

I fixed that issue by changing your proposal as follows:

```diff
--- a/lib/aur-graph
+++ b/lib/aur-graph
@@ -40,11 +40,9 @@ function setopt(opt, default) {
        return default
 }
 
-function has_valid_arch(attr,    i) {
-    if(i = index(attr, "_") && !(substr(attr, i+1) in arch))
-        return 0
-
-    return 1
+function has_valid_arch(attr) {
+    i = index(attr, "_")
+    return !i || (substr(attr, i+1) in arch)
 }
 
 BEGIN {
```

I also added a directive for shellcheck, which doesn’t know that `gawk` is a thing and would otherwise complain.

The script is now passing all my stress tests flawlessly.  
Thanks again for your contributions, which I think have improved the script.

claui commented:
@AladW All pending suggestions have been taken care of. I think this PR is now ready to merge.

Earnestly commented:
Nice, although I hoped to made the edit sooner, I forgot to include `delete(arch)` in the `$1 == "pkgbase"` pattern branch that would otherwise reset the arch array for subsequent pkgbases.

See edit in https://github.com/aurutils/aurutils/pull/1134#issuecomment-1922119914

claui commented:
@Earnestly Updated and re-tested.

Earnestly commented:
Bah, @AladW pointed out that the `$1 ~ /^provides/` check should also be `index($1, "provides") == 1`. With that fixed I don't see any issues with it. Thank you for testing these changes on your dataset, that has been very helpful.

AladW commented:
Yeah, I'm fine with either as long as it's consistent. Thank you both very much for all the work and analysis so far.

claui commented:
Fixed, thanks for the pointer!
